evm_trace/base.py

```diff
@@ -54,3 +57,121 @@ class CallTreeNode(BaseModel):
     @validator("address", "calldata", "returndata", pre=True)
     def validate_hexbytes(cls, v) -> HexBytes:
         return _convert_hexbytes(cls, v)
+
+
+def get_calltree_from_trace(
+    trace: Iterator[TraceFrame], show_internal=False, **root_node_kwargs
+) -> CallTreeNode:
+    """
+    Creates a CallTreeNode from a given transaction trace.
+
+    Args:
+        trace (Iterator[TraceFrame]): Iterator of transaction trace frames.
+        show_internal (bool): Boolean whether to display internal calls. Defaulted to False.
+        root_node_kwargs (dict): Keyword argments passed to the root ``CallTreeNode``.
+
+    Returns:
+        :class:`~evm_trace.base.CallTreeNode: Call tree of transaction trace.
+    """
+
+    return _create_node_from_call(
+        trace=trace,
+        show_internal=show_internal,
+        **root_node_kwargs,
+    )
+
+
+def _extract_memory(offset: HexBytes, size: HexBytes, memory: List[HexBytes]) -> HexBytes:
+    """
+    Extracts memory from the EVM stack.
+
+    Args:
+        offset (HexBytes): Offset byte location in memory.
+        size (HexBytes): Number of bytes to return.
+        memory (List[HexBytes]): Memory stack.
+
+    Returns:
+        HexBytes: Byte value from memory stack.
+    """
+
+    size_int = to_int(size)
+
+    if size_int == 0:
+        return HexBytes(b"")
+
+    offset_int = to_int(offset)
+
+    # Compute the word that contains the first byte
+    start_word = math.floor(offset_int / 32)
+    # Compute the word that contains the last byte
+    stop_word = math.ceil((offset_int + size_int) / 32)
+
+    byte_slice = b""
+
+    for word in memory[start_word:stop_word]:
+        byte_slice += word
+
+    offset_index = offset_int % 32
+    return HexBytes(byte_slice[offset_index:size_int])
+
+
+def _create_node_from_call(
+    trace: Iterator[TraceFrame], show_internal: bool = False, **node_kwargs
+) -> CallTreeNode:
+
+    if show_internal:
+        raise NotImplementedError()
+
+    node = CallTreeNode(**node_kwargs)
+
+    for frame in trace:
+        if frame.op in ("CALL", "DELEGATECALL", "STATICCALL"):
+            child_node_kwargs = {}
+
+            child_node_kwargs["address"] = frame.stack[-2][-20:]  # address is 20 bytes in EVM
+            child_node_kwargs["depth"] = frame.depth
+            # TODO: Validate gas values
+            child_node_kwargs["gas_limit"] = int(frame.stack[-1].hex(), 16)
+            child_node_kwargs["gas_cost"] = frame.gas_cost
+
+            if frame.op == "CALL":
+                child_node_kwargs["call_type"] = CallType.MUTABLE
+                child_node_kwargs["value"] = int(frame.stack[-3].hex(), 16)
+                child_node_kwargs["calldata"] = _extract_memory(
+                    offset=frame.stack[-4], size=frame.stack[-5], memory=frame.memory
```

Comment:
I am going to take your word that this is all doing what it needs to be doing!
A reference to the official docs describing the domain knowledge here would be nice.

Reply:
mostly from here: https://www.evm.codes/

Reply:
Added a link to the evm codes [f2cd2ef](https://github.com/ApeWorX/evm-trace/pull/4/commits/f2cd2efc98dbc46003082628584b66e7132b309d)